### PR TITLE
[RISCV] KASAN implementation.

### PR DESCRIPTION
**Pull request by MichalBlk**
Implementation of machine-dependent KASAN components.

### CODE REVIEW

File: sys/riscv/boot.c [hunk 1]
```diff
@@ -63,21 +65,30 @@
...
+
+#if KASAN
+#define BOOT_KASAN_SANITIZED_SIZE                                              \
+  roundup2(roundup2(KERNEL_VIRT_IMG_END, GROWKERNEL_STRIDE) +                  \
+             (VM_PAGE_PDS * GROWKERNEL_STRIDE) - KASAN_SANITIZED_START,        \
+           PAGESIZE * KASAN_SHADOW_SCALE_SIZE)
+
+#define BOOT_KASAN_SHADOW_SIZE                                                 \
+  (BOOT_KASAN_SANITIZED_SIZE / KASAN_SHADOW_SCALE_SIZE)
+#endif /* !KASAN */
```
Comment (pj1031999) [inline sys/riscv/boot.c:88]: Is that condition needed here? Could you just create unused definition?

**Review by pj1031999 — APPROVED** [review]
LGTM

CC: +jpszczolowski@

**Comment by cahirwpz** [comment]
@jpszczolowski Will you have time this week to have a look at this PR?

File: include/riscv/vm_param.h [hunk 2]
```diff
@@ -11,9 +11,12 @@
 #define USER_STACK_TOP 0x0000003fffff0000L
 #define USER_STACK_SIZE 0x800000 /* grows down up to that size limit */
 
+/* KASAN shadow map */
+#define KASAN_SHADOW_START 0xffffffd800000000L
+#define KASAN_MAX_SHADOW_SIZE (0xcL << 30) /* 12 GB */
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply include/riscv/vm_param.h]: I changed it to 8 as we'll run Mimiker on (emulated) HiFive Unleashed.

File: include/riscv/vm_param.h [hunk 3]
```diff
@@ -11,9 +11,12 @@
...
+
 #define VM_PAGE_PDS 32
 #define KSTACK_PAGES 2
-#define BOOTMEM_PAGES 40
 #else
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply include/riscv/vm_param.h]: Done.

File: sys/riscv/boot.c (hunk 1 above)
Comment (MichalBlk) [reply sys/riscv/boot.c]: Good point. This is the same remark as @cahirwpz has regarding the FPU- and SUM-related defines. Our preferred solution is not to enclose defines unnecessarily (even if they are unused with a specific configuration).

File: build/arch.riscv.mk [hunk 4]
```diff
@@ -29,4 +29,11 @@ ifeq ($(KERNEL), 1)
 	CFLAGS += -mcmodel=medany
 	CPPFLAGS += -DKERNEL_PHYS=$(KERNEL_PHYS)
 	CPPLDSCRIPT := 1
+	ifeq ($(KASAN), 1)
+	CFLAGS_KASAN = -fsanitize=kernel-address \
+		       -fasan-shadow-offset=0x90000000 \
+		       --param asan-globals=1 \
+		       --param asan-stack=1 \
+		       --param asan-instrument-allocas=1
+	endif
```
[the remark this answers is not in the document]
Comment (cahirwpz) [reply build/arch.riscv.mk:38]: I'll remember to add this information in #1270

**Review by cahirwpz — APPROVED** [review]

**Pull request merged**